tools/packaging/scripts/configure-hypervisor.sh

```diff
@@ -48,13 +48,27 @@ recognised_tags=(
 	[speed]="maximise startup speed"
 )
 
+# Wrapper around the bc command.
+# Return 0 if bc expression $1 is true, otherwise 1.
+calc() {
+	local expr="$1"
+	[ "$(echo $expr | bc)" == "1" ]
+}
+
 # Display message to stderr and exit indicating script failed.
 die() {
 	local msg="$*"
 	echo >&2 "$script_name: ERROR: $msg"
 	exit 1
 }
 
+# Compare with the QEMU version.
+# Use a bc expression as $1.
+qemu_version_is() {
+	local expr="$1"
+	calc "${qemu_version} $expr"
```

Comment:
Initially I was passing the `qemu_version` to `calc()` [1] as you mentioned above, @jodh-intel . But as I was repeating the same thing many times, I created the `qemu_version_is()`. I really don't have a strong preference here.

Also, before, the script was using local variables (`qemu_minor_version` and `qemu_major_version`) inside the method which were expected to be called from `main()`. So I really didn't change the semantic. I am also okay to promote `qemu_version`  global so easily accessible from anywhere in the script.

[1] btw, if you have a better name for this method then let me know

Reply:
@jodh-intel @devimc Would the following change be enough?

```
diff --git a/tools/packaging/scripts/configure-hypervisor.sh b/tools/packaging/scripts/configure-hypervisor.sh
index ba87b7e7..b2350aac 100755
--- a/tools/packaging/scripts/configure-hypervisor.sh
+++ b/tools/packaging/scripts/configure-hypervisor.sh
@@ -38,6 +38,9 @@ typeset -A recognised_tags
 # Prefix were kata will be installed
 prefix=${PREFIX:-/usr}
 
+# The QEMU version on "major.minor" format.
+qemu_version=""
+
 recognised_tags=(
        [arch]="architecture-specific"
        [functionality]="required functionality"
@@ -55,13 +58,6 @@ calc() {
        [ "$(echo $expr | bc)" == "1" ]
 }
 
-# Display message to stderr and exit indicating script failed.
-die() {
-       local msg="$*"
-       echo >&2 "$script_name: ERROR: $msg"
-       exit 1
-}
-
 # Compare with the QEMU version.
 # Use a bc expression as $1.
 qemu_version_is() {
@@ -221,7 +217,7 @@ generate_qemu_options() {
 
        # Disabled options
 
-       if qemu_version_is ">= 5" ; then
+       if calc "${qemu_version} >= 5" ; then
                # Disable sheepdog block driver support
                qemu_options+=(size:--disable-sheepdog)
 
@@ -255,7 +251,7 @@ generate_qemu_options() {
 
        # Disable PAM authentication: it's a feature used together with VNC access
        # that's not used. See QEMU commit 8953caf for more details
-       qemu_version_is ">= 4" && qemu_options+=(size:--disable-auth-pam)
+       calc "${qemu_version} >= 4" && qemu_options+=(size:--disable-auth-pam)
 
        # Disable unused filesystem support
        [ "$arch" == x86_64 ] && qemu_options+=(size:--disable-fdt)
@@ -264,7 +260,7 @@ generate_qemu_options() {
        qemu_options+=(size:--disable-libnfs)
 
        # Starting from QEMU 4.1, libssh replaces to libssh2
-       if qemu_version_is ">= 4.1" ; then
+       if calc "${qemu_version} >= 4.1" ; then
                qemu_options+=(size:--disable-libssh)
        else
                qemu_options+=(size:--disable-libssh2)
@@ -297,7 +293,7 @@ generate_qemu_options() {
 
        # SECURITY: Don't build a static binary (lowers security)
        # needed if qemu version is less than 2.7
-       if qemu_version_is "< 2.7" ; then
+       if calc "${qemu_version} < 2.7" ; then
                qemu_options+=(security:--disable-static)
        fi
 
@@ -340,7 +336,7 @@ generate_qemu_options() {
        # But since QEMU 5.2 the daemon is built as part of the tools set
        # (disabled with --disable-tools) thus it needs to be explicitely
        # enabled.
-       if qemu_version_is ">= 5.2" ; then
+       if calc "${qemu_version} >= 5.2" ; then
                qemu_options+=(functionality:--enable-virtiofsd)
        fi
 
@@ -370,12 +366,12 @@ generate_qemu_options() {
        # Disable Capstone
        qemu_options+=(size:--disable-capstone)
 
-       if qemu_version_is ">= 3" ; then
+       if calc "${qemu_version} >= 3" ; then
                # Disable graphics
                qemu_options+=(size:--disable-virglrenderer)
 
                # Due to qemu commit 3ebb9c4f52, we can't disable replication in v3.0
-               if qemu_version_is "> 3" ; then
+               if calc "${qemu_version} > 3" ; then
                        # Disable block replication
                        qemu_options+=(size:--disable-replication)
                fi
@@ -399,7 +395,7 @@ generate_qemu_options() {
 
                # vxhs was deprecated on QEMU 5.1 so it doesn't need to be
                # explicitly disabled.
-               if qemu_version_is "< 5.1" ; then
+               if calc "${qemu_version} < 5.1" ; then
                        qemu_options+=(size:--disable-vxhs)
                fi
        fi
@@ -416,7 +412,7 @@ generate_qemu_options() {
 
        # Always strip binaries
        # needed if qemu version is less than 2.7
-       if qemu_version_is "< 2.7" ; then
+       if calc "${qemu_version} < 2.7" ; then
                qemu_options+=(size:--enable-strip)
        fi
 
@@ -433,12 +429,12 @@ generate_qemu_options() {
        qemu_options+=(functionality:--enable-cap-ng)
        qemu_options+=(functionality:--enable-seccomp)
 
-       if qemu_version_is ">= 3.1" ; then
+       if calc "${qemu_version} >= 3.1" ; then
                # AVX2 is enabled by default by x86_64, make sure it's enabled only
                # for that architecture
                if [ "$arch" == x86_64 ]; then
                        qemu_options+=(speed:--enable-avx2)
-                       if qemu_version_is ">= 5" ; then
+                       if calc "${qemu_version} >= 5" ; then
                                qemu_options+=(speed:--enable-avx512f)
                        fi
                        # According to QEMU's nvdimm documentation: When 'pmem' is 'on' and QEMU is
@@ -471,7 +467,7 @@ generate_qemu_options() {
        _qemu_cflags=""
 
        # compile with high level of optimisation
-       if qemu_version_is "< 5.2" ; then
+       if calc "${qemu_version} < 5.2" ; then
                _qemu_cflags+=" -O3"
        fi
```

Reply:
can you apply and push it? I think it needs more changes, for example `local qemu_version=` -> `qemu_version=` in `main`

Reply:
@devimc good catch. also I deleted the `die()` method by mistake. Anyway, I just pushed another version of the commit.

Reply:
@devimc @jodh-intel , guys, I will drop the `bc` idea as it doesn't handle cases like '2.2 > 2.10' (return true, but should be false). I will send another implementation, where each digit is compared.

Reply:
good catch, try with `sort -V -C`

Reply:
@devimc now I understood when @jodh-intel said you had a trick with `sort` :)

What do you think about this:
```
# Given $1 and $2 as version strings with 'x.y.z' format; if $1 >= $2 then
# return 0. Otherwise return 1.
# Use this function on conditionals to compare versions.
#
gt_eq() {
        format='^[0-9]+(\.[0-9]+)*$'
        if [[ ! ("$1" =~ $format && "$2" =~ $format) ]]; then
                echo "ERROR: Malformed version string"
        fi
        echo -e "$1\n$2" | sort -V -r -C
}
```